README.md
# Subtitles Editor

This editor allows you to make edits to files with subtitles - shift the timestamps, do a linear correction, or translate them to another language.

## Supported File Formats

- SubRip Text (`.srt`)
- Web Video Text Tracks (`.vtt`)
- SubViewer (`.sbv`, `.sub`)

## Features

- Colorizer for subtitle files.
- Shift all timestamps by a specific time value.
- Renumber frame sequence with specified start index.
- Reorder frames based on their sequence number or timestamp.
- Linear correction by mapping two existing time stamps to new values.
- Translate subtitles to different languages.
- Only subtitles in selection are affected. If there is no selection, all subtitles are affected.

## Usage

Use command **Subtitles: Shift** and enter offset time to shift subtitles. The format is "HH:mm:ss,zzz". You can use negative time if subtitles are too late.

With **Subtitles: Renumber**, you can renumber the frames using the sequence with specified start index. After selecting this option, enter the start of the sequence (usually `1`) and it will be renumbered accordingly.

The command **Subtitles: Linear Correction** prompts you for two timestamp mappings. The input is in form "original timestamp -> new timestamp". Pick one point from the beginning of the movie and second one from the end of the movie to get the best approximation.

The command **Subtitles: Convert Time Format** converts timestamps in the file into one of the supported formats you can choose from (SRT, VTT, and SBV). Converting changes separator between start and end timestamp, separator for milliseconds, and number of digits used for milliseconds. You can also use this command to normalize the format of the timestamps if the file contains mixed timestamp formats.

Use command **Subtitles: Translate** to translate subtitles to another language using Google translation service.

With **Subtitles: Reorder by Sequence**, you can reorder the frames based on their sequence number. This can be useful if you want to work with translated and original subtitles at the same time. You can first translate the subtitles (which will replace the original ones) and append the original subtitles at the end. Then, you can reorder them so you will have translated and original frames near each other.

The **Subtitles: Reorder by Timestamp** reorders the frames by the time offset.

## For Developers

The goal of this extension is to allow easily modify subtitles files.

Tenet of this software is to do minimal modifications to the file when fulfilling any command, ignoring any errors that might be in the subtitles file. For instance, when asked to renumber frames, it should not modify the time format even if it is invalid. When altering timestamps, keep the time formatting.

All commands are to be applied to selection only. If no selection exists, they should be applied to whole file.

### Existing Subtitle Formats

There is bunch of [existing file formats](https://en.wikipedia.org/wiki/Category:Subtitle_file_formats). This extension currently supports SRT, VTT, and SBV files.

#### SubRip Text (SRT)

~~~srt
1
00:02:17,440 --> 00:02:20,375
Senator, we're making
our final approach into Coruscant.

2
00:02:20,476 --> 00:02:22,501 X1:100 X2:100 Y1:100 Y2:100
Very good, Lieutenant.
~~~

#### WebVTT (VTT)

~~~vtt
WEBVTT Kind: captions; Language: en

00:09.000 --> 00:11.000
<v Roger Bingham>We are in New York City

00:32.500 --> 00:33.500 align:start size:50%
<v Neil deGrasse Tyson><i>Laughs</i>
~~~

#### SubViewer (SBV, SUB)

~~~sbv
00:01:00.10,00:02:00.20
Oh, no.
The eggs are hatching!
~~~

### Updating Supported Languages

The languages that can be used for translation are published by Google in Google documentation. The extension uses languages defined in `src/languages.json` file.

To convert the published table to JSON:

- copy the table from Google documentation (https://cloud.google.com/translate/docs/languages)
- replace `^(.*?)\t(.*?)(?: or .*?)?(?: \(.+\))?$` with `\t"$2": "$1",`
- remove the last comma
- wrap the result in curly braces
- save it as `src/languages.json`

### Packaging

~~~
npm run package
# verify that packaged .vsix file contains the required files only
npm run publish
~~~

## Attributions

- [SubRip Text](https://en.wikipedia.org/wiki/SubRip) information by [Wikipedia Contributors](https://en.wikipedia.org/w/index.php?title=SubRip&action=history) is licensed under [CC BY-SA 3.0](https://creativecommons.org/licenses/by-sa/3.0/).
- [Web Video Text Tracks Format](https://developer.mozilla.org/en-US/docs/Web/API/WebVTT_API) information by [Mozilla Contributors](https://wiki.developer.mozilla.org/en-US/docs/Web/API/WebVTT_API$history) is licensed under [CC BY-SA 2.5](https://creativecommons.org/licenses/by-sa/2.5/).
- [SubViewer](https://wiki.videolan.org/SubViewer/) information by [VideoLAN Wiki Contributors](https://wiki.videolan.org/index.php?title=SubViewer&action=history) is licensed under [GPLv3](https://www.gnu.org/licenses/gpl-3.0.html).
- [Icon](https://commons.wikimedia.org/wiki/File:Font_Awesome_5_solid_closed-captioning.svg) from Font Awesome Free 5.2.0 by [@fontawesome](https://fontawesome.com) is licensed under [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/deed.en) and was modified to add white background.


CHANGELOG.md
# Change Log

All notable changes to the "subtitles-editor" extension are documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [1.1.6] - 2021-07-13

- Added option to sort the frames by timestamp.

## [1.1.6] - 2021-07-13

- Added option to format timestamp format.

## [1.1.5] - 2021-05-19

- Fix translation: Skip sequence and empty lines when translating.
- Update supported languages.
- Update dependencies and migrate from superseded module for running tests.

## [1.1.4] - 2021-04-10

- Fix translation response decoding.
- Fix detection of lines to not translate.

## [1.1.3] - 2020-12-16

- Take URI-encoding into calculation of text length that is sent to translation service.

## [1.1.2] - 2020-11-23

- Updated description.

## [1.1.1] - 2020-11-23

- Added icon and fixed README.

## [1.1.0] - 2020-11-23

- Added support for WebVTT and SubViewer formats.
- Use better namespace for commands.

## [1.0.4] - 2019-05-10

- Removed unnecessary files from the packaged extension.

## [1.0.3] - 2019-05-05

- Added option to reorder the frames based on their sequence number.

## [1.0.2] - 2019-02-08

- Updated dependencies.

## [1.0.1] - 2018-09-06

- Added option to renumber the sequence of the frames.

## [1.0.0] - 2018-08-18

- Added option to translate subtitles.

## [0.0.1] - 2018-07-08

- Initial release


src/test/index.ts
import * as path from 'path';
import * as Mocha from 'mocha';
import * as glob from 'glob';

export function run(): Promise<void> {
	// Create the mocha test
	const mocha = new Mocha({
		ui: 'tdd',
		color: true
	});

	const testsRoot = path.resolve(__dirname, '..');

	return new Promise((resolve, reject) => {
		glob('**/**.test.js', { cwd: testsRoot }, (err, files) => {
			if (err) {
				return reject(err);
			}

			// Add files to the test suite
			files.forEach(f => mocha.addFile(path.resolve(testsRoot, f)));

			try {
				// Run the mocha test
				mocha.run(failures => {
					if (failures > 0) {
						reject(new Error(`${failures} tests failed.`));
					} else {
						resolve();
					}
				});
			} catch (err) {
				console.error(err);
				reject(err);
			}
		});
	});
}


src/test/extension.test.ts

import * as assert from 'assert';

// import * as vscode from 'vscode';
// import * as myExtension from '../extension';

suite("Extension Tests", function () {
	test("Something 1", function() {
		assert.equal(-1, [1, 2, 3].indexOf(5));
		assert.equal(-1, [1, 2, 3].indexOf(0));
	});
});


src/test/model/Time.test.ts
import * as assert from 'assert';
import { Time } from '../../model/Time';
import { TimeLine } from '../../model/TimeLine';

function format(time: string): string {
	return Time.parse(time).format();
}

function shift(time: string, offset: string): string {
	const value = Time.parse(time);
	value.shift(Time.parse(offset).value);
	return value.format();
}

function applyLinearCorrection(time: string, originalTimeLine: string, updatedTimeLine: string): string {
	const value = Time.parse(time);
	const originalValue = TimeLine.parse(originalTimeLine);
	const updatedValue = TimeLine.parse(updatedTimeLine);
	if (originalValue && updatedValue) {
		value.applyLinearCorrection(originalValue, updatedValue);
	}
	return value.format();
}

function convert(timeLine: string, timeSeparator: string, millisSeparator: string, shortMillis: boolean): string {
	const value = TimeLine.parse(timeLine);
	if (!value) {
		return timeLine;
	}
	value.convert(timeSeparator, millisSeparator, shortMillis);
	return value.format();
}

suite('Time Tests', function() {
	test('format', function() {
		assert.strictEqual(format('12:34:56,7891'), '12:34:56,789');
		assert.strictEqual(format('12:34:56.7891'), '12:34:56.789');
		assert.strictEqual(format('12:34:56,789'), '12:34:56,789');
		assert.strictEqual(format('12:34:56.789'), '12:34:56.789');
		assert.strictEqual(format('12:34:56,78'), '12:34:56,78');
		assert.strictEqual(format('12:34:56.78'), '12:34:56.78');
		assert.strictEqual(format('12:34:56,7'), '12:34:56,700');
		assert.strictEqual(format('12:34:56.7'), '12:34:56.700');
		assert.strictEqual(format('34:56,789'), '00:34:56,789');
		assert.strictEqual(format('34:56.789'), '00:34:56.789');
		assert.strictEqual(format('34:56,78'), '00:34:56,78');
		assert.strictEqual(format('34:56.78'), '00:34:56.78');
		assert.strictEqual(format('-12:34:56,7891'), '-12:34:56,789');
		assert.strictEqual(format('-12:34:56.7891'), '-12:34:56.789');
		assert.strictEqual(format('-12:34:56,789'), '-12:34:56,789');
		assert.strictEqual(format('-12:34:56.789'), '-12:34:56.789');
		assert.strictEqual(format('-12:34:56,78'), '-12:34:56,78');
		assert.strictEqual(format('-12:34:56.78'), '-12:34:56.78');
		assert.strictEqual(format('-12:34:56,7'), '-12:34:56,700');
		assert.strictEqual(format('-12:34:56.7'), '-12:34:56.700');
		assert.strictEqual(format('-34:56,789'), '-00:34:56,789');
		assert.strictEqual(format('-34:56.789'), '-00:34:56.789');
		assert.strictEqual(format('-34:56,78'), '-00:34:56,78');
		assert.strictEqual(format('-34:56.78'), '-00:34:56.78');
		assert.strictEqual(format('+12:34:56,7891'), '12:34:56,789');
		assert.strictEqual(format('+12:34:56.7891'), '12:34:56.789');
		assert.strictEqual(format('+12:34:56,789'), '12:34:56,789');
		assert.strictEqual(format('+12:34:56.789'), '12:34:56.789');
		assert.strictEqual(format('+12:34:56,78'), '12:34:56,78');
		assert.strictEqual(format('+12:34:56.78'), '12:34:56.78');
		assert.strictEqual(format('+12:34:56,7'), '12:34:56,700');
		assert.strictEqual(format('+12:34:56.7'), '12:34:56.700');
		assert.strictEqual(format('+34:56,789'), '00:34:56,789');
		assert.strictEqual(format('+34:56.789'), '00:34:56.789');
		assert.strictEqual(format('+34:56,78'), '00:34:56,78');
		assert.strictEqual(format('+34:56.78'), '00:34:56.78');
	});

	test('shift', function() {
		assert.strictEqual(shift('65:43:21,123', '11:11:11,111'), '76:54:32,234');
		assert.strictEqual(shift('01:00:00,000', '-02:00:00,000'), '-01:00:00,000');
	});

	test('linear correction', function() {
		assert.strictEqual(applyLinearCorrection('01:00:00,000', '00:00:00,000-->02:00:00,000', '01:00:00,000-->05:00:00,000'), '03:00:00,000');
	});

	test('convert', function() {
		assert.strictEqual(convert('01:23:45,678-->23:45:12,345', ' --> ', '.', false), '01:23:45.678 --> 23:45:12.345');
		assert.strictEqual(convert('01:23:45,678-->23:45:12,345', ',', '.', true), '01:23:45.68,23:45:12.35');

		assert.strictEqual(convert('11:23:45.678-->23:45:12.345', ' --> ', ',', false), '11:23:45,678 --> 23:45:12,345');
		assert.strictEqual(convert('11:23:45.678-->23:45:12.345', ',', '.', true), '11:23:45.68,23:45:12.35');

		assert.strictEqual(convert('21:23:45.68,23:45:12.35', ' --> ', ',', false), '21:23:45,680 --> 23:45:12,350');
		assert.strictEqual(convert('21:23:45.68,23:45:12.35', ' --> ', '.', false), '21:23:45.680 --> 23:45:12.350');
	});
});


.eslintrc.js
module.exports = {
	"env": {
		"es6": true,
		"node": true
	},
	"parser": "@typescript-eslint/parser",
	"parserOptions": {
		"project": "tsconfig.json",
		"sourceType": "module"
	},
	"plugins": [
		"@typescript-eslint"
	],
	"rules": {
		"@typescript-eslint/member-delimiter-style": [
			"warn",
			{
				"multiline": {
					"delimiter": "semi",
					"requireLast": true
				},
				"singleline": {
					"delimiter": "semi",
					"requireLast": false
				}
			}
		],
		"@typescript-eslint/naming-convention": "warn",
		"@typescript-eslint/no-unused-expressions": "warn",
		"@typescript-eslint/semi": [
			"warn",
			"always"
		],
		"indent": ["error", "tab", { "SwitchCase": 1 }],
		"curly": "warn",
		"eqeqeq": [
			"warn",
			"always"
		],
		"no-trailing-spaces": "error",
		"no-throw-literal": "warn"
	}
};


src/model/Time.ts
import { TimeLine } from "./TimeLine";

export const TIME_REGEX = /^\s*(?<sign>[+\-]?)(?:(?<hours>\d+):)?(?<minutes>\d+):(?<seconds>\d{2})(?:(?<separator>[,.])(?<millis>\d+))?\s*$/;
export const TIME_TAG_REGEX = /\<((?<sign>[+\-]?)(?:(?<hours>\d+):)?(?<minutes>\d+):(?<seconds>\d{2})(?:(?<separator>[,.])(?<millis>\d+))?)\>/g;

export class Time {
	constructor(
		public value: number,
		public separator: string,
		public shortMillis: boolean,
	) {
	}

	shift(offset: number): void {
		this.value += offset;
	}

	applyLinearCorrection(original: TimeLine, updated: TimeLine): void {
		const factor = (updated.endTime.value - updated.startTime.value)
			/ (original.endTime.value - original.startTime.value);
		this.value = (this.value - original.startTime.value) * factor + updated.startTime.value;
	}

	convert(separator: string, shortMillis: boolean) {
		this.separator = separator;
		this.shortMillis = shortMillis;
	}

	normalize(): void {
		this.separator = ',';
		this.shortMillis = false;
	}

	static value(sign: string, hours: string, minutes: string, seconds: string, millis: string): number {
		return (sign === '-' ? -1 : 1) * (Number(hours || 0) * 60 * 60
			+ Number(minutes || 0) * 60
			+ Number(seconds || 0)
			+ Number((millis || '').substr(0, 3)) / Time.pow10(String(millis || '').length));
	}

	static parse(text: string): Time {
		const groups = text?.match(TIME_REGEX)?.groups || {};
		const value = Time.value(groups.sign, groups.hours, groups.minutes, groups.seconds, groups.millis);
		const separator = groups.separator || ',';
		const shortMillis = (groups.millis || '').length === 2;
		return new Time(value, separator, shortMillis);
	}

	format(): string {
		let value = Math.abs(this.value);
		const hours = Math.floor(value / (60 * 60));
		value -= hours * 60 * 60;
		const minutes = Math.floor(value / 60);
		value -= minutes * 60;
		const seconds = Math.floor(value);
		value -= seconds;
		const millis = Math.round(value * (this.shortMillis ? 100 : 1000));
		return (this.value < 0 ? '-' : '')
			+ Time.pad2(hours)
			+ ':'
			+ Time.pad2(minutes)
			+ ':'
			+ Time.pad2(seconds)
			+ this.separator
			+ (this.shortMillis ? Time.pad2(millis) : Time.pad3(millis));
	}

	private static pow10(value: number): number {
		switch (value) {
			case 0: return 1;
			case 1: return 10;
			case 2: return 100;
			default:
			case 3: return 1000;
		}
	}

	private static pad2(value: number): string {
		return value < 10 ? '0' + String(value) : String(value);
	}

	private static pad3(value: number): string {
		if (value < 10) {
			return '00' + String(value);
		} else if (value < 100) {
			return '0' + String(value);
		} else {
			return String(value);
		}
	}
}


src/model/Frame.ts
import * as vscode from 'vscode';
import { TimeLine } from './TimeLine';

export interface Frame {
	lineIndex: number;
	sequence: number;
	timeLine: TimeLine | null;
	lines: vscode.TextLine[];
}

export namespace Frame {
	export function compareSequence(a: Frame, b: Frame): number {
		return a.sequence !== b.sequence ? a.sequence - b.sequence : a.lineIndex - b.lineIndex;
	}

	export function compareTimestamp(a: Frame, b: Frame): number {
		if (a.timeLine !== null || b.timeLine !== null) {
			if (a.timeLine === null) {
				return -1;
			}
			if (b.timeLine === null) {
				return 1;
			}
			if (a.timeLine.startTime !== b.timeLine.startTime) {
				return a.timeLine.startTime.value - b.timeLine.startTime.value;
			}
			if (a.timeLine.endTime !== b.timeLine.endTime) {
				return a.timeLine.endTime.value - b.timeLine.endTime.value;
			}
		}
		return compareSequence(a, b);
	}
}


src/extension.ts
'use strict';

import * as vscode from 'vscode';
import * as https from 'https';
import * as languages from './languages.json';
import { URL } from 'url';
import { Frame } from './model/Frame';
import { TimeLine } from './model/TimeLine';
import { Time, TIME_REGEX, TIME_TAG_REGEX } from './model/Time';

const sequenceRegex = /^([+\-]?)(\d+)(?=\s|$)/;
const timeMappingRegex = /^\d{2}:\d{2}:\d{2}[,.]\d{3} -> \d{2}:\d{2}:\d{2}[,.]\d{3}$/;

async function shift() {
	const textEditor = vscode.window.activeTextEditor;

	if (typeof textEditor === 'undefined') {
		return false;
	}

	const inputBox = {
		placeHolder: 'Time shift',
		prompt: 'Enter the desired time shift. Use negative value if subtitles are late.',
		value: '00:00:00,000',
		valueSelection: [6, 8] as [number, number],
		validateInput: async (value: string) => TIME_REGEX.test(value) ? null : 'Time has to be in format ±00:00:00,000.'
	};

	const value = await vscode.window.showInputBox(inputBox);

	if (typeof value === 'undefined') {
		return false;
	}

	const offset = Time.parse(value).value;
	const workspaceEdit = new vscode.WorkspaceEdit();
	const documentUri = textEditor.document.uri;
	const selections = !textEditor.selection.isEmpty
		? textEditor.selections
		: [new vscode.Selection(textEditor.document.positionAt(0), textEditor.document.lineAt(textEditor.document.lineCount - 1).range.end)];

	for (const selection of selections) {
		for (let lineIndex = selection.start.line; lineIndex <= selection.end.line; ++lineIndex) {
			const line = textEditor.document.lineAt(lineIndex);
			if (!line.isEmptyOrWhitespace) {
				const timeLine = TimeLine.parse(line.text);
				if (timeLine) {
					timeLine.shift(offset);
					workspaceEdit.replace(documentUri, line.range, timeLine.format());
				} else if (TIME_TAG_REGEX.test(line.text)) {
					workspaceEdit.replace(documentUri, line.range,
						line.text.replace(TIME_TAG_REGEX, (_match, text) => {
							const time = Time.parse(text);
							time.shift(offset);
							return `<${time.format()}>`;
						}));
				}
			}
		}
	}

	await vscode.workspace.applyEdit(workspaceEdit);

	return true;
}

async function renumber() {
	const textEditor = vscode.window.activeTextEditor;

	if (typeof textEditor === 'undefined') {
		return false;
	}

	const inputBox = {
		placeHolder: 'Sequence start index',
		prompt: 'Enter the start index to renumber the sequence.',
		value: '1',
		valueSelection: [0, 1] as [number, number],
		validateInput: async (value: string) => sequenceRegex.test(value) ? null : 'Offset has to be a number.'
	};

	const value = await vscode.window.showInputBox(inputBox);

	if (typeof value === 'undefined') {
		return false;
	}

	let offset = Number(value);
	const workspaceEdit = new vscode.WorkspaceEdit();
	const documentUri = textEditor.document.uri;
	const selections = !textEditor.selection.isEmpty
		? textEditor.selections
		: [new vscode.Selection(textEditor.document.positionAt(0), textEditor.document.lineAt(textEditor.document.lineCount - 1).range.end)];

	for (const selection of selections) {
		for (let lineIndex = selection.start.line; lineIndex <= selection.end.line; ++lineIndex) {
			if (isEmptyLine(textEditor.document, lineIndex - 1) && isSequenceLine(textEditor.document, lineIndex)) {
				const line = textEditor.document.lineAt(lineIndex);
				workspaceEdit.replace(documentUri, line.range, String(offset) + line.text.replace(sequenceRegex, ''));
				++offset;
			}
		}
	}

	await vscode.workspace.applyEdit(workspaceEdit);

	return true;
}

async function reorderBySequence() {
	return await reorder(Frame.compareSequence);
}

async function reorderByTimestamp() {
	return await reorder(Frame.compareTimestamp);
}

async function reorder(comparer: (a: Frame, b: Frame) => number) {
	const textEditor = vscode.window.activeTextEditor;

	if (typeof textEditor === 'undefined') {
		return false;
	}

	const workspaceEdit = new vscode.WorkspaceEdit();
	const textDocument = textEditor.document;
	const documentUri = textDocument.uri;
	const selections = !textEditor.selection.isEmpty
		? textEditor.selections
		: [new vscode.Selection(textDocument.positionAt(0), textDocument.lineAt(textDocument.lineCount - 1).range.end)];

	for (const selection of selections) {
		const frames: Frame[] = [];
		let frame: Frame | null = null;
		for (let lineIndex = selection.start.line; lineIndex <= selection.end.line; ++lineIndex) {
			const line = textDocument.lineAt(lineIndex);
			const previousLineIsEmpty = isEmptyLine(textDocument, lineIndex - 1);
			let timeLine = null;
			if (previousLineIsEmpty && isSequenceLine(textDocument, lineIndex)) {
				frame = {
					lineIndex,
					sequence: Number.parseInt(line.text, 10),
					timeLine,
					lines: [line]
				};
				frames.push(frame);
			} else if (previousLineIsEmpty && (timeLine = TimeLine.parse(line.text))) {
				frame = {
					lineIndex,
					sequence: Number.NEGATIVE_INFINITY,
					timeLine,
					lines: [line]
				};
				frames.push(frame);
			} else {
				if (!frame) {
					frame = {
						lineIndex,
						sequence: Number.NEGATIVE_INFINITY,
						timeLine,
						lines: []
					};
					frames.push(frame);
				}
				if (!frame.timeLine) {
					frame.timeLine = TimeLine.parse(line.text);
				}
				frame.lines.push(line);
			}
		}
		frames.sort(comparer);
		const lines = frames.reduce((acc, frame) => { acc.push(...frame.lines.map(line => line.text + '\n')); return acc; }, [] as string[]);
		workspaceEdit.replace(documentUri, selection, lines.join('').replace(/\n$/, ''));
	}

	await vscode.workspace.applyEdit(workspaceEdit);

	return true;
}

function isEmptyLine(textDocument: vscode.TextDocument, lineIndex: number): boolean {
	return lineIndex < 0 || textDocument.lineAt(lineIndex).isEmptyOrWhitespace;
}

function isSequenceLine(textDocument: vscode.TextDocument, lineIndex: number): boolean {
	return sequenceRegex.test(textDocument.lineAt(lineIndex).text);
}

function findFirstTime(textDocument: vscode.TextDocument, lineNumbers: number[]): Time {
	for (const lineNumber of lineNumbers) {
		const line = textDocument.lineAt(lineNumber);
		const timeLine = TimeLine.parse(line.text);
		if (timeLine) {
			timeLine.startTime.normalize();
			return timeLine.startTime;
		}
	}
	return Time.parse('');
}

async function linearCorrection() {
	const textEditor = vscode.window.activeTextEditor;

	if (typeof textEditor === 'undefined') {
		return false;
	}

	const keys = Array.from(Array(textEditor.document.lineCount).keys());
	const firstTime = findFirstTime(textEditor.document, keys).format();
	const secondTime = findFirstTime(textEditor.document, keys.reverse()).format();

	const firstInputBox = {
		placeHolder: 'Time #1',
		prompt: 'Enter first time for subtitle that appears early in the video in format old time -> correct time.',
		value: `${firstTime} -> ${firstTime}`,
		valueSelection: [16, 28] as [number, number],
		validateInput: async (value: string) => timeMappingRegex.test(value) ? null : 'Time has to be in format 00:00:00,000 -> 00:00:00,000.'
	};

	const secondInputBox = {
		...firstInputBox,
		placeHolder: 'Time #2',
		prompt: 'Enter second time for subtitle that appears late in the video in format old time -> correct time.',
		value: `${secondTime} -> ${secondTime}`,
	};

	const firstValue = await vscode.window.showInputBox(firstInputBox);
	if (typeof firstValue === 'undefined') {
		return false;
	}

	const secondValue = await vscode.window.showInputBox(secondInputBox);
	if (typeof secondValue === 'undefined') {
		return false;
	}

	const firstMapping = firstValue.split(' -> ').map(Time.parse);
	const secondMapping = secondValue.split(' -> ').map(Time.parse);
	const originalTimeLine = new TimeLine(firstMapping[0], secondMapping[0]);
	const updatedTimeLine = new TimeLine(firstMapping[1], secondMapping[1]);

	const workspaceEdit = new vscode.WorkspaceEdit();
	const documentUri = textEditor.document.uri;
	const selections = !textEditor.selection.isEmpty
		? textEditor.selections
		: [new vscode.Selection(textEditor.document.positionAt(0), textEditor.document.lineAt(textEditor.document.lineCount - 1).range.end)];

	for (const selection of selections) {
		for (let lineIndex = selection.start.line; lineIndex <= selection.end.line; ++lineIndex) {
			const line = textEditor.document.lineAt(lineIndex);
			const timeLine = TimeLine.parse(line.text);
			if (timeLine) {
				timeLine.applyLinearCorrection(originalTimeLine, updatedTimeLine);
				workspaceEdit.replace(documentUri, line.range, timeLine.format());
			} else if (TIME_TAG_REGEX.test(line.text)) {
				workspaceEdit.replace(documentUri, line.range,
					line.text.replace(TIME_TAG_REGEX, (_match, text) => {
						const time = Time.parse(text);
						time.applyLinearCorrection(originalTimeLine, updatedTimeLine);
						return `<${time.format()}>`;
					}));
			}
		}
	}

	await vscode.workspace.applyEdit(workspaceEdit);

	return true;
}

interface SubtitleType {
	name: string;
	extensions: string[];
	timeSeparator: string;
	millisSeparator: string;
	shortMillis: boolean;
}

async function convertTimeFormat() {
	const textEditor = vscode.window.activeTextEditor;

	if (typeof textEditor === 'undefined') {
		return false;
	}

	const subtitleTypes: { [name: string]: SubtitleType } = {
		['SubRip Text']: {
			name: 'SubRip Text',
			extensions: ['srt'],
			timeSeparator: ' --> ',
			millisSeparator: ',',
			shortMillis: false,
		},
		['Web Video Text Tracks']: {
			name: 'Web Video Text Tracks',
			extensions: ['vtt'],
			timeSeparator: ' --> ',
			millisSeparator: '.',
			shortMillis: false,
		},
		['SubViewer']: {
			name: 'SubViewer',
			extensions: ['sbv', 'sub'],
			timeSeparator: ',',
			millisSeparator: '.',
			shortMillis: true,
		},
	};
	const items = Object.values(subtitleTypes)
		.map(({ name, extensions }) => ({ label: name, detail: extensions.join(', ') }));

	const quickPickOpts: vscode.QuickPickOptions = {
		placeHolder: 'Type',
		matchOnDetail: true
	};

	const value = await vscode.window.showQuickPick(items, quickPickOpts);

	if (typeof value === 'undefined') {
		return false;
	}

	const subtitleType = subtitleTypes[value.label];
	const workspaceEdit = new vscode.WorkspaceEdit();
	const documentUri = textEditor.document.uri;
	const selections = !textEditor.selection.isEmpty
		? textEditor.selections
		: [new vscode.Selection(textEditor.document.positionAt(0), textEditor.document.lineAt(textEditor.document.lineCount - 1).range.end)];

	for (const selection of selections) {
		for (let lineIndex = selection.start.line; lineIndex <= selection.end.line; ++lineIndex) {
			const line = textEditor.document.lineAt(lineIndex);
			if (!line.isEmptyOrWhitespace) {
				const timeLine = TimeLine.parse(line.text);
				if (timeLine) {
					timeLine.convert(subtitleType.timeSeparator, subtitleType.millisSeparator, subtitleType.shortMillis);
					workspaceEdit.replace(documentUri, line.range, timeLine.format());
				}
			}
		}
	}

	await vscode.workspace.applyEdit(workspaceEdit);

	return true;
}

async function translate() {
	const textEditor = vscode.window.activeTextEditor;

	if (typeof textEditor === 'undefined') {
		return false;
	}
	const languagesByCode = languages as { [code: string]: string };
	const items = Object.keys(languagesByCode)
		.map(code => ({
			label: `${languagesByCode[code]}`,
			detail: code
		}));

	const quickPickOpts: vscode.QuickPickOptions = {
		placeHolder: 'Language',
		matchOnDetail: true
	};

	const value = await vscode.window.showQuickPick(items, quickPickOpts);

	if (typeof value === 'undefined') {
		return false;
	}

	const language = value.detail;

	const workspaceEdit = new vscode.WorkspaceEdit();
	const documentUri = textEditor.document.uri;
	const selections = !textEditor.selection.isEmpty
		? textEditor.selections
		: [new vscode.Selection(textEditor.document.positionAt(0), textEditor.document.lineAt(textEditor.document.lineCount - 1).range.end)];

	const originalLines: string[] = [];
	const lineIndexes: number[] = [];
	for (const selection of selections) {
		for (let lineIndex = selection.start.line; lineIndex <= selection.end.line; ++lineIndex) {
			const line = textEditor.document.lineAt(lineIndex);
			if (!line.isEmptyOrWhitespace && !TimeLine.parse(line.text) && !isSequenceLine(textEditor.document, lineIndex)) {
				originalLines.push(line.text);
				lineIndexes.push(lineIndex);
			}
		}
	}

	const translatedLines = await translateLines(language, originalLines);
	while (true) {
		const lineIndex = lineIndexes.shift();
		const translatedLine = translatedLines.shift();
		if (lineIndex === undefined || translatedLine === undefined) {
			break;
		}
		const line = textEditor.document.lineAt(lineIndex);
		workspaceEdit.replace(documentUri, line.range, translatedLine);
	}

	await vscode.workspace.applyEdit(workspaceEdit);

	return true;
}

async function translateText(language: string, text: string): Promise<string> {
	const url = `https://translate.googleapis.com/translate_a/single?client=gtx&sl=auto&tl=${encodeURIComponent(language)}&dt=t&q=${encodeURIComponent(text)}`;
	const json = JSON.parse(await httpGet(url));
	return json[0].map((x: any[]) => x[0]).join('');
}

async function translateLines(language: string, originalLines: string[]): Promise<string[]> {
	const result: string[] = [];
	const lines: string[] = [];
	let length = 0;
	for (const originalLine of originalLines) {
		const originalLineLength = encodeURIComponent(originalLine).length;
		if (length + originalLineLength + 1 > 8000) {
			const translatedText = await translateText(language, lines.join('\n'));
			Array.prototype.push.apply(result, translatedText.split('\n'));
			lines.length = 0;
			length = 0;
		}
		lines.push(originalLine);
		length += originalLineLength + 1;
	}
	const translatedText = await translateText(language, lines.join('\n'));
	Array.prototype.push.apply(result, translatedText.split('\n'));
	return result.map(x => x.replace(/(<)(\/?)\s*([bi])\s*(>)/gi, '$1$2$3$4'));
}

async function httpGet(url: string): Promise<string> {
	return new Promise<string>((resolve, reject) => {
		const uri = new URL(url);
		const opts = {
			hostname: uri.hostname,
			port: uri.port || 443,
			path: uri.pathname + uri.search,
			headers: {
				// eslint-disable-next-line @typescript-eslint/naming-convention
				'User-Agent': 'Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/49.0.2623.110 Safari/537.36'
			}
		};
		https.get(opts, res => {
			if (res.statusCode !== 200) {
				reject(new Error(``));
			}
			let body = '';
			res.setEncoding('utf8');
			res.on('data', d => body += d);
			res.on('end', () => resolve(body));
		});
	});
}

export function activate(context: vscode.ExtensionContext) {
	context.subscriptions.push(vscode.commands.registerCommand('subtitles.shift', shift));
	context.subscriptions.push(vscode.commands.registerCommand('subtitles.renumber', renumber));
	context.subscriptions.push(vscode.commands.registerCommand('subtitles.reorderBySequence', reorderBySequence));
	context.subscriptions.push(vscode.commands.registerCommand('subtitles.reorderByTimestamp', reorderByTimestamp));
	context.subscriptions.push(vscode.commands.registerCommand('subtitles.linearCorrection', linearCorrection));
	context.subscriptions.push(vscode.commands.registerCommand('subtitles.convertTimeFormat', convertTimeFormat));
	context.subscriptions.push(vscode.commands.registerCommand('subtitles.translate', translate));
}

export function deactivate() {
}
